File: leet/1306_jump_game_3/1306_jump_game_3.py
# LeetCode 1306. Jump Game III
# https://leetcode.com/problems/jump-game-iii/
# O(n) time O(1) space
class Solution:
    def canReach(self, arr: List[int], start: int) -> bool:
        def dfs(i):
            if i < 0 or i >= len(arr) or arr[i] < 0:
                return False
            if arr[i] == 0:
                return True
            arr[i] *= -1
            return dfs(i - arr[i]) or dfs(i + arr[i])
        return dfs(start)

File: leet/1306_jump_game_3/1306_jump_game_3.ts
// LeetCode 1306. Jump Game III
// https://leetcode.com/problems/jump-game-iii/
// O(n) time-and-space
function canReach(arr: number[], start: number): boolean {
    const visited = new Array(arr.length).fill(false);
    const dfs = (i: number): boolean => {
        if (i < 0 || i >= arr.length || visited[i]) return false;
        if (arr[i] == 0) return true;
        visited[i] = true;
        return dfs(i + arr[i]) || dfs(i - arr[i]);
    }
    return dfs(start);
};

